### ISSUE TITLE
Simplify argument type checks and exception raise 

### ISSUE THREAD

**Issue by srcansiz** [fedbiomed/fedbiomed#878]
Some of argument type checks seems to me not necessary. These type checks are usually never reached within the library unless developer does mistake while coding. Therefore, I propose to remove some type checking for the methods, functions  that does not rely on user input. However, developer should make sure that his/her implementation is correct. Using a tool like `mypy` to validate method argument types can help developer to make sure type definitions and implementation is correct.   

For example: 

- Private methods are not exposed to the user. The arguments there are passed to the private methods are coded by the developer. If the argument type is wrong this is a developer mistake rather than user input. (unless private method uses a user input).  Therefore, I propose to remove type checks for private methods if it does not rely on user input and make sure that it is functional before merging (with unit testing).  

- Another example is `TrainRequest` or other request that are send from researcher to node or vice versa. The types of these arguments are already validated in the messaging layer.  For example, checking if `researacher_id` is `str` or not is not relevant. It is also same for `TrainingArguments`. 

- In Fed-BioMed, each class basically has its own exception. However, sometimes, the exception that does not make sense to developer. For example, when a  `RoundError` is raised due to AttributeError somewhere in the library (developer mistake), it is not easy to trace back. Therefore, some exceptions should  stay as it is unless it highly depends on user input. For example, `training_step` function of  `TrainingPlan`.

**Comment by sharkovsky** [comment]
I fully agree 100%! 

I have a couple more opinions:
- for exceptions, I think there used to be an interest in converting everything into a Fed-BioMed exception because these could be treated differently by the runtime. For example, we may decide that the node CLI should crash if a non-Fedbiomed exception is raised (because we don't know what to expect), but if in fedbiomed exception was raised we might assume that at least it is a known issue/possibility and therefore no need to crash the node. However, I think that it would make more sense to raise a `Fedbiomed AttributeError` in your example, rather than what we currently do. 
- I would be in favour of strongly reducing the types of check we do on `TrainingArguments` (or even more extrem, I would be in favour of removing them altogether)

**Comment by mvesin** [comment]
Hi @srcansiz,

For me, this kind of things about agreeing on coding practices (or changing the existing ones) should be discussed and decided together, maybe in a retrospective meeting (or dedicated meeting ?). And then written and followed :-)

At first sight, I would tend to agree with 1. but
disagree with 2. and 3. that we purposely implemented for robustness.

**Comment by srcansiz** [comment]
Hi @sharkovsky, @mvesin, 

Thanks for your comments. Yes I agree @mvesin It can be something that we can discuss during the retrospective or  we can have an internal meeting during this sprint. The main idea is to reduce the complexity of the code, avoid writing unnecessary code, and thanks to that reducing the time spent while coding, unit-testing etc.

**Comment by mvesin** [comment]
If matters of coding rules (these ones, maybe others) are of interest to the team, maybe we can dedicate next sprint retrospective to discussing & deciding on some of these aspects ?

**Comment by srcansiz** [comment]
Yes, I agree it matters of coding rules even though we don't have specific written rule for that.  My point is that it seems we won't have enough time discuss and decide it during retrospective. However, I think we will need maximum 1 hour to discuss and decide, we can spare 1 hour when everyone is in SED. But if you wanna delay it for the next sprint yes we can.

**Comment by mvesin** [comment]
I agree we have few coding rules, but exceptions is one of the few really specified and documented aspects: https://github.com/fedbiomed/fedbiomed/blob/develop/CODING_RULES.md

But it doesn't seem we apply these rules (eg I am the first one to use `except Exceptions:`) ...
It probably means we need to agree on evolved rules.

**Comment by leonard2103** [comment]
mypy implementation is continued in issue #1762
